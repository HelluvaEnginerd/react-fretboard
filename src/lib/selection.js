/**
 * helper functions that return commonly used selection objects.
 */

import { contains } from 'ramda'
import { Distance, Interval, Chord, Scale } from 'tonal'

export const intervalNotes = (note, ivl) =>
  [
    note,
    Distance.transpose(note, Interval.fromSemitones(ivl)),
  ]

const namedIntervalNote = (tonic, nameStatus) => note => ({
  note,
  status: nameStatus ? Distance.interval(tonic, note) : 'selected',
  label: Distance.interval(tonic, note),
})

export const namedIntervalNotes = (note, ivl, nameStatus = false) =>
  intervalNotes(note, ivl).map(namedIntervalNote(note, nameStatus))

export const chordNotes = chord =>
  Chord.notes(chord)

const chordTonic = chord =>
  Chord.tokenize(chord)[0]

export const namedChordNotes = (chord, nameStatus = false) =>
  chordNotes(chord).map(namedIntervalNote(chordTonic(chord), nameStatus))

export const scaleNotes = (tonic, scale) =>
  Scale.notes(tonic, scale)

export const namedScaleNotes = (tonic, scale, nameStatus = false) =>
  scaleNotes(tonic, scale).map(namedIntervalNote(tonic, nameStatus))

const position = (openNote, width, note) => {
  const dist = Distance.semitones(openNote, note)
  return width > dist && dist >= 0
    ? dist
    : null
}

// experimental
export const triadShape = (tuning, width, chord, str) => {
  const notes = chordNotes(chord)
  const positions = notes.map((note, i) =>
    position(tuning[(tuning.length - (str + 1)) + i], width, note))
  const chordExists = !contains(null, positions)

  return chordExists
    ? positions.map((pos, i) => ({ str: str - i, pos }))
    : null
}
