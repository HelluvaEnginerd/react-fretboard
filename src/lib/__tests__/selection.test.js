import { intervalNotes } from '../selection'

it('should return array of interval pcs', () => {
  const note = 'C'
  const ivl = 4 // 3M
  const expected = ['C', 'E']
  expect(intervalNotes(note, ivl)).toEqual(expected)
})

it('should return array of interval pitches', () => {
  const note = 'C4'
  const ivl = 4 // 3M
  const expected = ['C4', 'E4']
  expect(intervalNotes(note, ivl)).toEqual(expected)
})
