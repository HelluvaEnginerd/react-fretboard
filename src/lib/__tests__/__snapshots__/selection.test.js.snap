// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selection, chord shapes should return array locs 1`] = `
Array [
  Object {
    "pos": 5,
    "str": 2,
  },
  Object {
    "pos": 4,
    "str": 1,
  },
  Object {
    "pos": 3,
    "str": 0,
  },
]
`;

exports[`selection, scales named pc noteObj 1`] = `
Array [
  Object {
    "label": "1P",
    "note": "C4",
    "status": "selected",
  },
  Object {
    "label": "2M",
    "note": "D4",
    "status": "selected",
  },
  Object {
    "label": "3M",
    "note": "E4",
    "status": "selected",
  },
  Object {
    "label": "4P",
    "note": "F4",
    "status": "selected",
  },
  Object {
    "label": "5P",
    "note": "G4",
    "status": "selected",
  },
  Object {
    "label": "6M",
    "note": "A4",
    "status": "selected",
  },
  Object {
    "label": "7M",
    "note": "B4",
    "status": "selected",
  },
]
`;

exports[`selection, scales should return array of scale pcs 1`] = `
Array [
  "C",
  "D",
  "E",
  "F",
  "G",
  "A",
  "B",
]
`;

exports[`selection, scales should return array of scale pitches 1`] = `
Array [
  "C4",
  "D4",
  "E4",
  "F4",
  "G4",
  "A4",
  "B4",
]
`;
